java-checks/src/main/java/org/sonar/java/checks/CollectionMethodsWithLinearComplexityCheck.java

```diff
@@ -114,7 +114,16 @@ private static boolean isField(Symbol symbol) {
   }
 
   private static boolean matchesActualType(Symbol invocationTarget, Set<String> actualTypes) {
-    return actualTypes.contains(actualType(invocationTarget).fullyQualifiedName());
+    Type declaredType = invocationTarget.type();
+    if (actualTypes.contains(declaredType.fullyQualifiedName())) {
+      return true;
+    }
+    if (invocationTarget.isPrivate() || invocationTarget.isFinal()) {
```

Comment:
This part worth a comment to explain _why_ we focus only on private or final fields (type cannot be changed from the outside)